ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/ClientTest.java

```diff
@@ -350,29 +337,66 @@ private void verifyPullQueryServerState() {
     assertThat(server.getQueryIDs(), hasSize(0));
   }
 
-  private void shouldDeliver(
+  private static void shouldReceiveRows(
       final Publisher<Row> publisher,
-      final int numRows,
       final boolean subscriberCompleted
   ) {
     TestSubscriber<Row> subscriber = new TestSubscriber<Row>() {
       @Override
       public synchronized void onSubscribe(final Subscription sub) {
         super.onSubscribe(sub);
-        sub.request(numRows);
+        sub.request(DEFAULT_JSON_ROWS.size());
       }
     };
     publisher.subscribe(subscriber);
-    assertThatEventually(subscriber::getValues, hasSize(numRows));
-    for (int i = 0; i < numRows; i++) {
-      assertThat(subscriber.getValues().get(i).values(), equalTo(rowWithIndexAsKsqlArray(i)));
-    }
+    assertThatEventually(subscriber::getValues, hasSize(DEFAULT_JSON_ROWS.size()));
+
+    verifyRows(subscriber.getValues());
+
     assertThatEventually(subscriber::isCompleted, equalTo(subscriberCompleted));
     assertThat(subscriber.getError(), is(nullValue()));
   }
 
-  private static KsqlArray rowWithIndexAsKsqlArray(final int index) {
-    return new KsqlArray(rowWithIndex(index).getList());
+  private static void verifyRows(final List<Row> rows) {
+    assertThat(rows, hasSize(DEFAULT_JSON_ROWS.size()));
+    for (int i = 0; i < DEFAULT_JSON_ROWS.size(); i++) {
+      verifyRowWithIndex(rows.get(i), i);
+    }
+  }
+
+  private static void verifyRowWithIndex(final Row row, final int index) {
+    // verify metadata
+    assertThat(row.values(), equalTo(EXPECTED_ROWS.get(index)));
+    assertThat(row.columnNames(), equalTo(DEFAULT_COLUMN_NAMES));
+    assertThat(row.columnTypes(), equalTo(DEFAULT_COLUMN_TYPES));
+
+    // verify type-based getters
+    assertThat(row.getString("f_str"), is("foo" + index));
+    assertThat(row.getInt("f_int"), is(index));
+    assertThat(row.getBoolean("f_bool"), is(index % 2 == 0));
+    assertThat(row.getLong("f_long"), is(Long.valueOf(index * index)));
+    assertThat(row.getDouble("f_double"), is(index + 0.1111));
+    assertThat(row.getDecimal("f_decimal"), is(BigDecimal.valueOf(index + 0.1)));
+    assertThat(row.getKsqlArray("f_array"), is(new KsqlArray().add("s" + index).add("t" + index)));
```

Comment:
I think we should also test the API of KsqlObject and KsqlArray here (or in another test case). I.e. check all the methods on those have been exercised.

Reply:
I'm not a fan of adding comprehensive testing to KsqlObject and KsqlArray here since it feels unnecessary (a unit test should suffice) but I've gone ahead and added it anyway.

Incidentally, adding more assertions here revealed a bug in PollableSubscriber (completely unrelated to KsqlObject and KsqlArray) so I'm grateful for having done the exercise at your behest ;)

Reply:
Imho, as a general rule, anything that is accessible via the API should be tested via the API. Testing only using isolated fine grained unit tests can often not expose issues that only manifest when using that part of the API like a user would. To me those are the really important tests to have.

In terms of terminology - I wouldn't say ClientTest is an integration test. Unfortunately the terminology around testing is quite poor, with "unit test" often used to mean small, fine grained tests, and integration tests where the unit is the entire system. But in reality the "unit" size is a spectrum and can vary from between these two extremes. Often the most useful kinds of tests (e.g. this ClientTest) are where the unit size is in the middle. Here the unit size is the client as a whole.

I do still think we need an integration test. By integration test I don't mean ClientTest - I consider that a coarse grained unit test. But a test that tests the client with the real(ish) server (like RestApiTest) and tests sending some real queries to it :)

Reply:
Yeah, I've been unsure whether to call ClientTest a unit test or an integration test. If only there were another term in between, haha.

Rest assured when I've been saying "integration tests coming in a future PR" I've been using that to mean your understanding of an integration test -- testing the client with TestKsqlRestApp :)

Reply:
Maybe coarse-grained unit test, or module test?